### ISSUE TITLE
DOLFINx 3rd Party Apps

### ISSUE THREAD

**Issue by jhale** [FEniCS/dolfinx#2397]
It would be really useful to have a list of community-supported third-party add-ons for DOLFINx. I think this is especially important as DOLFINx is relatively minimal, and consequently some higher-level functionality will be provided by community packages.

An issue in the past is that third-party add-ons often become unmaintained, and so the utility of the list becomes questionable.

However, I think with a modern CI approach we could bypass this issue.

My idea is as follows:

- A repository i.e. `FEniCS/community-packages`.
- Contains a `README.md` that is also built and pushed to GitHub pages.
- The main information in the `README.md` is a table containing the following columns:
     - The name of the package (sorted by this column) as a link to the homepage/github/gitlab page.
     - A short description limited to a certain number of characters.
     - A stable CI run badge (see below for details).
     - A main CI run badge (see below for details).

To be included, the package must:

- Use DOLFINx.
- Have unit tests.
- Have demos.
- Present a CI pipeline running weekly: last stable tag against `dolfinx/dolfinx:stable`.
- Present a CI pipeline running weekly: `main` branch against `dolfinx/dolfinx:nightly`.

Inclusion in the list will be made by third-party package maintainers submitting a PR to `FEniCS/community-packages`.

**Comment by francesco-ballarin** [comment]
I would be very happy with this proposal, and I could contribute at least two packages ;)

My main concern is about:
* Present a CI pipeline running weekly: last stable tag against `dolfinx/dolfinx:stable`.

At least in my case (but that might quite common ;) ), I do not have a big team of people working with me on these projects, so by asking me to support both `main` and `stable` you are (sort of) doubling the amount of work I have to put in. Right now, with the limited time I have, and thanks a nightly CI run on upstream `dolfinx/dolfinx:main`, I can only manage to support `main`.

**Comment by jhale** [comment]
Completely understand, many of my packages are in the same situation.

The point of the page is to provide good information to users, rather than to make developers feel awkward about not having enough time to keep up. Perhaps this could be clarified in `README.md`.

I also want to re-iterate that we would not expecting that your `main` branch stays simultaneously compatible with our `stable` and 'nightly'. The stable CI badge says that single tag (of your creation/choosing) stays compatible with `stable`. Then it easy for users to see if the tag will work with versions from conda, Debian etc.

**Comment by jhale** [comment]
Another good option would be to simply have the option to replace the first badge with e.g. 'no stable versions`.

**Comment by francesco-ballarin** [comment]
> Another good option would be to simply have the option to replace the first badge with e.g. 'no stable versions`.

Or tell if the package targets compatibility with main, stable or both.

> The stable CI badge says that single tag (of your creation/choosing) stays compatible with stable.

1. That's too strong an assumption. You would be forcing downstream users to tag versions in sync with dolfinx tagged versions. It's my package, I'll surely be free of tagging a release whenever I feel like ;)
2. What would be the point of having a weekly CI run if both upstream and downstream are tagged? Both are fixed targets, so it's not like if something breaks the downstream developer could fix it, unless they create a new tag. This part of the proposal would make better sense IMHO if the downstream target was a branch, but as per my previous comment that would mean doubling the work for the downstream team.

**Comment by francesco-ballarin** [comment]
(sorry, maybe my point of view in 2 is not clear. The case I have in mind is: as a downstream developer adds a new feature, but in that the new feature there is call to a `dolfinx` functionality for which the signature changed between `stable` and `main`.)

**Comment by jhale** [comment]
> Or tell if the package targets compatibility with main, stable or both.

We are trying to create a resource with a primary goal of giving users quality information. In my experience that "compatibility pledges" are worthless. The only thing that asserts compatibility is proof, and a two weekly CI runs and the badges give concrete information about the current state of a package.

Regarding your other points:

1. A key question users have is "does a tag (or git ref) of this package work with the latest stable version of DOLFINx?". The stable CI run badge answers this question.

2. None, I agree. But upstream is not fixed, we control `:stable` and it moves. The developer controls their tag (or git ref, if you prefer) and they can also move it.

In short, I think a fair compromise is that you can assert that your package does not assert compatibility with main or stable. But to be included on the list it must have one, and that you must demonstrate it with the CI run.

**Comment by jhale** [comment]
Modified version.

My idea is as follows:

- A repository i.e. `FEniCS/community-packages`.
- Contains a `README.md` that is also built and pushed to GitHub pages.
- The main information in the `README.md` is a table containing the following columns:
     - The name of the package (sorted by this column) as a link to the homepage/github/gitlab page.
     - A short description limited to a certain number of characters.
     - One or both of:
         - A stable CI run badge (see below for details). If not present, "No guaranteed compatibility".
         - A main CI run badge (see below for details). If not present, "No guaranteed compatibility".

To be included, the package must:

- Use DOLFINx.
- Have unit tests.
- Have demos.
- _Stable CI badge:_ Present a CI pipeline running weekly: a git ref/tag/branch (specified by the developer in their CI file) against `dolfinx/dolfinx:stable` (controlled by FEniCS Steering Council).
- _Main CI badge:_ Present a CI pipeline running weekly: `main` branch against `dolfinx/dolfinx:nightly`.

Inclusion in the list will be made by third-party package maintainers submitting a PR to `FEniCS/community-packages`.

**Comment by francesco-ballarin** [comment]
> running weekly

I'll interpret this as weekly or even more frequently than that (e.g. daily) :)

**Comment by jhale** [comment]
Of course, you can also make it nightly if you wish! 

I'd be happy with allowing up to fortnightly for the `main` badge. I don't want to put developers off with the thought of fixing red CI badges every morning.

**Comment by jhale** [comment]
Modified version.

My idea is as follows:

- A repository i.e. `FEniCS/community-packages`.
- Contains a `README.md` that is also built and pushed to GitHub pages.
- The main information in the `README.md` is a table containing the following columns:
     - The name of the package (sorted by this column) as a link to the homepage/github/gitlab page.
     - A short description limited to a certain number of characters.
     - One or both of:
         - A stable CI run badge (see below for details). If not present, "Not guaranteed".
         - A main CI run badge (see below for details). If not present, "Not guaranteed".

To be included, the package must:

- Use DOLFINx.
- Have unit tests.
- Have demos.
- _Stable CI badge:_ Present a CI pipeline running on a **maximum fortnightly period**: a git ref/tag/branch (specified by the developer in their CI file) against `dolfinx/dolfinx:stable` (controlled by FEniCS Steering Council).
- _Main CI badge:_ Present a CI pipeline running weekly: `main` branch against `dolfinx/dolfinx:nightly`.

Inclusion in the list will be made by third-party package maintainers submitting a PR to `FEniCS/community-packages`.

**Comment by mscroggs** [comment]
This would be great.

If possible, it would be very nice to show: if the software no longer passes CI, the last version number that something used to work with

**Comment by garth-wells** [comment]
Principle agreed.
